htlcswitch/switch.go

```diff
@@ -0,0 +1,399 @@
+	default:
+		return errors.New("wrong request type")
+	}
+}
+
+// startHandling start handling inner command requests and print the
+// htlc switch statistics.
+// NOTE: Should be run as goroutine.
+func (s *HTLCSwitch) startHandling() {
+	defer s.service.Done()
+
+	// TODO(roasbeef): cleared vs settled distinction
+	var prevNumUpdates uint64
+	var prevSatSent btcutil.Amount
+	var prevSatRecv btcutil.Amount
+
+	statsTicker := time.NewTicker(10 * time.Second)
+
+	for {
+		select {
+		case command := <-s.commands:
+			command.err <- s.forwardRequest(command.request)
+
+		case <-statsTicker.C:
+			var overallNumUpdates uint64
+			var overallSatSent btcutil.Amount
+			var overallSatRecv btcutil.Amount
+
+			for _, manager := range s.managers {
+				overallNumUpdates += manager.NumUpdates()
+				overallSatSent += manager.SatSent()
+				overallSatRecv += manager.SatRecv()
+			}
+
+			if overallNumUpdates == 0 {
+				continue
+			}
+
+			diffNumUpdates := overallNumUpdates - prevNumUpdates
+			diffSatSent := overallSatSent - prevSatSent
+			diffSatRecv := overallSatRecv - prevSatRecv
+
+			log.Infof("Sent %v satoshis, received %v satoshi in "+
+				"the last 10 seconds (%v tx/sec)",
+				diffSatSent.ToUnit(btcutil.AmountSatoshi),
+				diffSatRecv.ToUnit(btcutil.AmountSatoshi),
+				float64(diffNumUpdates)/10)
+
+			prevNumUpdates = overallNumUpdates
+			prevSatSent = overallSatSent
+			prevSatRecv = overallSatRecv
+
+		case <-s.service.Quit:
+			return
+		}
+	}
+}
+
+// Start starts all helper goroutines required for the operation of the switch.
+func (s *HTLCSwitch) Start() error {
+	log.Infof("HTLC Switch starting")
+	if err := s.service.Start(); err != nil {
+		return err
+	}
+
+	s.service.Go(s.startHandling)
+	return nil
+}
+
+// Stop gracefully stops all active helper goroutines, then waits until they've
+// exited.
+func (s *HTLCSwitch) Stop() error {
+	log.Infof("HTLC Switch shutting down")
+	if err := s.service.Stop(nil); err != nil {
+		return err
+	}
+
+	for _, manager := range s.managers {
+		s.remove(manager)
+	}
+
+	return nil
+}
+
+// Add is used to add and start the newly created HTLC manager ans start use it
+// to propagate HTLCs.
+func (s *HTLCSwitch) Add(manager HTLCManager) error {
+	if err := manager.Start(); err != nil {
+		return err
+	}
+
+	s.mutex.Lock()
+	s.managers[*manager.ID()] = manager
+	s.mutex.Unlock()
+
+	log.Infof("Added htlc manager for ChannelPoint(%v), bandwidth=%v",
+		manager.ID(), manager.Bandwidth())
+	return nil
+}
+
+// Remove is used to remove and stop the htlc manager by channel point of the
+// channel which htlc manager is managing.
+func (s *HTLCSwitch) RemoveByChan(chanPoint *wire.OutPoint) error {
+	manager, err := s.Get(chanPoint)
+	if err != nil {
+		return err
+	}
+
+	return s.remove(manager)
+}
+
+// Get returns the htlc manager which corresponds to the channel which he is
+// managing.
+func (s *HTLCSwitch) Get(chanPoint *wire.OutPoint) (HTLCManager, error) {
```

Comment:
I don't think this method (nor many of the related methods need to be publicaly exported).

Also naming suggestion: `LookUpLink`.